lib_xua/src/core/clocking/sw_pll_wrapper.xc

```diff
@@ -106,87 +106,97 @@ void do_sw_pll_phase_frequency_detector_dig_rx( unsigned short mclk_time_stamp,
-        select
+        unsigned rx_word = 0;
+        while(running)
         {
-            case inuint_byref(c_sigma_delta, rx_word):
-                if(rx_word == DISABLE_SDM)
-                {
-                    f_error = 0;
-                    running = 0;
-                }
-                else
-                {
-                    f_error = (int32_t)rx_word;
-                    unsafe
+            /* Poll for new SDM control value */
+            select
+            {
+                case inuint_byref(c_sw_pll, rx_word):
+                    inct(c_sw_pll);
+                    if(rx_word == DISABLE_SDM)
                     {
-                        sw_pll_sdm_do_control_from_error(&sw_pll, -f_error);
-                        dco_setting = sw_pll.sdm_state.current_ctrl_val;
+                        f_error = 0;
+                        running = 0;
                     }
-                }
-            break;
-
-            /* Do nothing & fall-through. Above case polls only once per loop */
-            default:
-            break;
-        }
-
-        /* Wait until the timer value has been reached
-           This implements a timing barrier and keeps
-           the loop rate constant. */
-        select
-        {
-            case tmr when timerafter(time_trigger) :> int _:
-                time_trigger += sdm_interval;
-            break;
-        }
-
-        unsafe {
-            sw_pll_do_sigma_delta(&sw_pll.sdm_state, this_tile, dco_setting);
-        }
-    } /* if running */
+                    else
+                    {
+                        f_error = (int32_t)rx_word;
+                        unsafe
+                        {
+                            sw_pll_sdm_do_control_from_error(&sw_pll, -f_error);
+                            dco_setting = sw_pll.sdm_state.current_ctrl_val;
+                        }
+                    }
+                break;
+
+                /* Do nothing & fall-through. Above case polls only once per loop */
+                default:
+                break;
+            }
+
+            /* Wait until the timer value has been reached
+               This implements a timing barrier and keeps
+               the loop rate constant. */
+            select
+            {
+                case tmr when timerafter(time_trigger) :> int _:
+                    time_trigger += sdm_interval;
+                break;
+            }
+
+            unsafe {
+                sw_pll_do_sigma_delta(&sw_pll.sdm_state, this_tile, dco_setting);
+            }
+        } /* while running */
+    } /* while(1) */
 }
 
 
-void restart_sigma_delta(chanend c_sigma_delta)
+void restart_sigma_delta(chanend c_sw_pll, unsigned selected_mclk_rate)
 {
-    outuint(c_sigma_delta, DISABLE_SDM); /* Resets SDM */
+    outuint(c_sw_pll, DISABLE_SDM); /* Resets SDM */
+    outct(c_sw_pll, XS1_CT_END);
+    outuint(c_sw_pll, selected_mclk_rate);
+    outct(c_sw_pll, XS1_CT_END);
```

Comment:
if you want to completely remove the chance of dead lock here do:
outct()
inct()
outuint()
outct() (not sure if this is required but you had this one)
outuint()
outct()

Reply:
@xross I'd like to chat about this - I see you use the idiom I have used elsewhere in the code. I think what you are suggesting basically makes it c <: val; ?

Reply:
It's not quite  c <: val, that would generate a lot more operations. With the CT_END handshake you guarantee that the other end is ready to take a word. Since it's a CT_END you want clog up a route.  If you send a word blindly you could potentially deadlock if all routes are used up. If you require some decoupling then this doesn't really work.